system/jlib/jmetrics.hpp

```diff
@@ -141,6 +153,22 @@ class jlib_decl GaugeMetric : public Metric
     }
 };
 
+template<typename T>
+class CustomMetric : public MetricBase
+{
+public:
+    CustomMetric(const char *name, const char *desc, MetricType metricType, std::atomic<T> &_value) :
```

Comment:
My concern is synchronizing reading the value during a report (which is assured to be on a separate thread). This comes down to a decision on who implements the queryValue method. If the templated class does, then it needs to ensure synchronization with the component code updating the value. If not implemented by the template, then the component must provide a subclass and implement the method and provide whatever synchronization is needed. 

I chose to implement in the template class and force the template parameter to be an atomic, that way synchronization is handled. A component can still provide a full implementation of the IMetric interface.

I guess it comes down to whether synchronization of reading and updating the value should be guaranteed or not.

Reply:
There is no guaranteed synchronization between reading and updating of the metric values - they are always an arbitrary snapshot in time, and whether or not a value has been incremented depends on inter thread timings.

The question is whether reading the value will be atomic. (Will you get a value where half the bytes are form one value, and half from another?)  In practice on all modern processors reading a properly aligned 4 or 8 byte unsigned will be atomic.
Restricting this class to std::atomic means that it will not work if RelaxedAtomic is changed to make std::atomic a protected base class.  It is simpler, and more flexible, to allow any type.